File: Java/225-ImplementStackUsingQueues.java
class MyStack {
    Queue<Integer> q = new LinkedList<>();
    // Push element x onto stack.
    public void push(int x) {
        q.add(x);
        for (int i = 1; i < q.size(); i++)
            q.add(q.poll());
    }

    // Removes the element on top of the stack.
    public void pop() {
        q.remove();
    }

    // Get the top element.
    public int top() {
        return q.peek();
    }

    // Return whether the stack is empty.
    public boolean empty() {
        return q.isEmpty();
    }


    //with two queues
    private Queue<Integer> q1 = new LinkedList<>();
    private Queue<Integer> q2 = new LinkedList<>();

    public Integer peek() {
       return q1.isEmpty() ? q2.peek() : q1.peek();
    }

    public Integer pop() {
       if (!q1.isEmpty()){
           q1.poll();
       }else{
           q2.poll();
       }
    }

    public void push(Integer item) {
       if (q1.isEmpty()){
           q1.add(item);
           for (int i = 0; i < q2.size(); i++){
               q1.add(q2.poll());
           }
       }else{
           q2.add(item);
           for (int i = 0; i < q1.size(); i++){
               q2.add(q1.poll());
           }
       }
    }

    public boolean isEmpty() {
       return q1.isEmpty() && q2.isEmpty();
    }
}

File: cpp/225-ImplementStackUsingQueues.cpp
class MyStack {
public:
    queue<int> q1;
    queue<int> q2;

    /** Initialize your data structure here. */
    MyStack() {

    }

    /** Push element x onto stack. */
    void push(int x) {
        if (q1.empty()){
            q1.push(x);
            while (!q2.empty()){
                q1.push(q2.front());
                q2.pop();
            }
        }else{
            q2.push(x);
            while (!q1.empty()){
                q2.push(q1.front());
                q1.pop();
            }
        }
    }

    /** Removes the element on top of the stack and returns that element. */
    int pop() {
        if (!q1.empty()){
            int res = q1.front();
            q1.pop();
            return res;
        }else{
            int res = q2.front();
            q2.pop();
            return res;
        }
    }

    /** Get the top element. */
    int top() {
        return !q1.empty() ? q1.front() : q2.front();
    }

    /** Returns whether the stack is empty. */
    bool empty() {
        return q1.empty() && q2.empty();
    }
};

/**
 * Your MyStack object will be instantiated and called as such:
 * MyStack obj = new MyStack();
 * obj.push(x);
 * int param_2 = obj.pop();
 * int param_3 = obj.top();
 * bool param_4 = obj.empty();
 */

 //One queue:
 class MyStack {
public:
    queue<int> q;

    /** Initialize your data structure here. */
    MyStack() {

    }

    /** Push element x onto stack. */
    void push(int x) {
        q.push(x);
        for (int i = 1; i < q.size(); i++){
            int item = q.front();
            q.pop();
            q.push(item);
        }
    }

    /** Removes the element on top of the stack and returns that element. */
    int pop() {
        int item = q.front();
        q.pop();
        return item;
    }

    /** Get the top element. */
    int top() {
        return q.front();
    }

    /** Returns whether the stack is empty. */
    bool empty() {
        return q.empty();
    }
};

/**
 * Your MyStack object will be instantiated and called as such:
 * MyStack obj = new MyStack();
 * obj.push(x);
 * int param_2 = obj.pop();
 * int param_3 = obj.top();
 * bool param_4 = obj.empty();
 */
